[sBTC] Feat: Peg out request & fulfillment wire formats

netrome opened the pull request:
Identical to #3510 except this is pointing directly to `next`.

sergey-shandar commented on src/chainstate/burn/operations/test/mod.rs:
```diff
@@ -0,0 +1,91 @@
+mod serialization;
+
+pub(crate) fn seeded_rng() -> StdRng {
+    SeedableRng::from_seed([0; 32])
```
Awesome! IMHO, we need some determinism in the repo.

sergey-shandar reviewed (approved):
LGTM

jcnelson commented on src/chainstate/burn/operations/peg_out_request.rs:
```diff
@@ -0,0 +1,384 @@
+    }
+
+    #[test]
+    fn test_parse_peg_out_request_should_succeed_given_a_conforming_transaction_with_extra_memo_bytes(
```
Minor formatting nit -- please remove the newline.  `()` belongs on one line.

netrome replied:
Done :heavy_check_mark:

jcnelson commented on testnet/stacks-node/src/burnchains/bitcoin_regtest_controller.rs:
```diff
@@ -1214,6 +1221,152 @@ impl BitcoinRegtestController {
+        self.finalize_tx(
+            epoch_id,
+            &mut tx,
+            payload.amount + payload.fulfillment_fee,
```
This should just be `payload.fulfillment_fee`.  This transaction only spends BTC to pay for the pegout fulfillment fee.

netrome replied:
Done :heavy_check_mark:

jcnelson reviewed (changes requested):
Almost there.  Still have a wrong payment calculation.

netrome commented:
> Almost there. Still have a wrong payment calculation.

Interesting. I've rebased and updated the payment calculation now. As mentioned in the sBTC meeting, I will add the changes from https://github.com/stacks-network/stacks-blockchain/pull/3576 to this branch since that is such a small PR. Also, it logically makes sense to incorporate those changes here to get the full end-to-end implementation of the wire formats including RPC endpoints.

I'll post here when it's done. Just awaiting @stjepangolemac to finish an update to the RPC PR before I cherry-pick.

netrome commented:
This PR has now been updated to include the RPC changes. You can view the RPC changes in isolation [here](https://github.com/stacks-network/stacks-blockchain/pull/3577/files/1f5efe2ccd0f8b61eab1f23caafed16ffd69ad6d..beceaf6e93b388c202c45dff7fe3998fe6319b81)

@kantai and @jcnelson it would be great if you could have a look at this when you have time.

kantai commented on docs/rpc/api/core-node/get-burn-ops.example.json:
```diff
@@ -1,14 +1,41 @@
-      "vtxindex": 2
-    }
-  ]
+    "peg_in": [
+        {
+            "amount": 1337,
+            "block_height": 218,
+            "burn_header_hash": "3292a7d2a7e941499b5c0dcff2a5656c159010718450948a60c2be9e1c221dc4",
+            "memo": "",
+            "peg_wallet_address": "tb1pqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkgkkf5",
+            "recipient": "S0000000000000000000002AA028H.awesome_contract",
+            "txid": "d81bec73a0ea0bdcf9bc011f567944eb1eae5889bf002bf7ae641d7096157771",
+            "vtxindex": 2
+        }
+    ],
+    "peg_out_request": [
+        {
+            "amount": 1337,
+            "recipient": "1BixGeiRyKT7NTkJAHpWuP197KXUNqhCU9",
+            "block_height": 218,
+            "burn_header_hash": "3292a7d2a7e941499b5c0dcff2a5656c159010718450948a60c2be9e1c221dc4",
+            "peg_wallet_address": "tb1qqvpsxqcrqvpsxqcrqvpsxqcrqvpsxqcrqvpsxqcrqvpsxqcrqvps3f3cyq",
+            "txid": "d81bec73a0ea0bdcf9bc011f567944eb1eae5889bf002bf7ae641d7096157771",
+            "vtxindex": 2,
+            "signature": "0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d0d",
+            "fulfillment_fee": 0,
+            "memo": "00010203"
+        }
+    ],
+    "peg_out_fulfill": [
```
Is this example actually a possible return value for the RPC? I don't think it can contain `peg_out_fulfill` at the same time as `peg_out_request`. The OpenAPI spec can contain multiple examples, but I think they'd need to be in separate example files.

stjepangolemac replied:
I think we wanted to show all variations of the response. It seems there's a better way - add multiple examples to the spec.

https://www.viralpatel.net/openapi-multiple-examples/

netrome replied:
Thanks for highlighting this! Let's update to use separate examples to the spec then. @stjepangolemac will you do it or should I?

netrome replied:
I've updated it now :heavy_check_mark:

stjepangolemac replied:
Yeah I was trying to push but thought my git broke down. I didn't realize you pushed identical changes first so that confused me. Thanks!

kantai reviewed (approved):
This LGTM, just had a comment on the documentation example.

netrome commented:
@jcnelson Everything is green now, we've got two approvals and I've resolved recent conflicts. Feel free to merge if things look good to you.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
